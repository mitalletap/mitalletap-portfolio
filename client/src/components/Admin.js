import React, { Component } from 'react';
import TextField from '@material-ui/core/TextField';
import Grid from '@material-ui/core/Grid';
import Button from '@material-ui/core/Button';
import Snackbar from '@material-ui/core/Snackbar';
import Alert from '@material-ui/lab/Alert';
import S3 from 'react-aws-s3';

import '../App.css';

class Admin extends Component {
    constructor(props) {
      super(props);
      this.state = {
        fileName: '',
        fileType: '',
        fileData: [],
        projectName: '',
        githubUrl: '',
        demoUrl: '',
        image: '',
        description: '',
        clicked: false,
        completed: false,
        status: "",
        open: false
      }


    }

    componentDidMount()
 {
     
 }
    allFieldsComplete() {
        const { fileName, projectName, githubUrl, demoUrl, description, fileData } = this.state;
        return (githubUrl !== "" && projectName !== '' && demoUrl !== '' && description !== '' && fileData.size > 0);
    }

    handleProjectName = e => {
        this.setState({ projectName: e.target.value })
    } 
    
    handleFileName = e => {
        this.setState({ fileName: e.target.value })
    }

    handleGithub = e => {
        this.setState({ githubUrl: e.target.value })
        
    }

    handleDemo = e => {
        this.setState({ demoUrl: e.target.value })
        
    }

    handleDescription = e => {
        this.setState({ description: e.target.value })

    }

    handleFileSelected = e => {
        this.setState({ image: e.target.files[0].name, fileData: e.target.files[0] });
    }


    handleSubmit = () => {
        const { fileData, fileName } = this.state;
        const config = {
            bucketName: process.env.REACT_APP_BUCKET_NAME,
            region: process.env.REACT_APP_BUCKET_REGION,
            accessKeyId: process.env.REACT_APP_S3_KEY,
            secretAccessKey: process.env.REACT_APP_S3_SECRET,
            s3Url: `https://s3.${process.env.REACT_APP_BUCKET_REGION}.amazonaws.com/${process.env.REACT_APP_BUCKET_NAME}`
        }

        const Client = new S3(config);
        Client.uploadFile(fileData, fileName)
        .then((res) => this.setState({ status: "true", open: true }))
        .catch((err) => this.setState({ status: "false", open: true }));

        this.setState({ clicked: true, completed: true });
    }

    handleClose = () => {
        this.setState({
          ...this.state,
          open: false,
        });
      };

    render() { 
        const { image, projectNam, completed, open } = this.state;
        const horizontal = "center";
        const vertical = "top";
        return (  
            <div className="admin-page">
                <Snackbar open={open} autoHideDuration={1000} onClose={this.handleClose} anchorOrigin={{ vertical, horizontal }}>
                    {
                        this.state.status === "true" ?
                        <Alert severity="success">
                            Successfully saved Project!
                        </Alert> 
                        :
                        <Alert severity="error">
                            Failed to save Project!
                        </Alert>
                    }
                </Snackbar>
                <h1> Admin Page </h1>
                <div className="admin-grid">
                    <Grid container spacing={4}> 
                        <Grid item xs={12}>
                            <TextField id="user-program" size="small" label="Project Name" variant="outlined" onChange={this.handleProjectName}/>
                        </Grid>
                        <Grid item xs={12}>
                            <TextField id="user-file" size="small" label="File Name" variant="outlined" onChange={this.handleFileName}/>
                        </Grid>
                        <Grid item xs={12}>
                            <TextField id="user-github" size="small" label="Github URL" variant="outlined" onChange={this.handleGithub}/>
                        </Grid>
                        <Grid item xs={12}>
                            <TextField id="user-demo" size="small" label="Demo URL" variant="outlined" onChange={this.handleDemo}/>
                        </Grid>
                        <Grid item xs={12}>
                            <TextField id="user-description" size="small" label="Description" variant="outlined" onChange={this.handleDescription}/>
                        </Grid>
                        <Grid item xs={12}>
                            <input accept="image/*" id="icon-button-file" type="file" onChange={this.handleFileSelected}/>
                            <label htmlFor="icon-button-file">
                                <Button variant="contained" color="primary" component="span"> Upload Image </Button>
                            </label>
                        </Grid>
                        <Grid item xs={12}>
                            <TextField id="standard-multiline-static" InputProps={{ readOnly: true }} placeholder={image} rows={4} />
                        </Grid>
                        <Grid item xs={12}>
                                <Button disabled={!this.allFieldsComplete()} className="submit-button" variant="contained" color="primary" component="span" onClick={this.handleSubmit}> Submit Information </Button>
                        </Grid>
                    </Grid>
                </div>
            <button onClick={() => console.log(this.state)}> state </button>
            </div>
        );
    }
}
 
export default Admin;