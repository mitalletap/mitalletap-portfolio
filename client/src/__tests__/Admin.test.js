import React, { Component } from 'react';
import { shallow, mount } from 'enzyme';
import Admin from '../components/Admin';
import TextField from '@material-ui/core/TextField';
import fetchMock from 'fetch-mock';

describe('tests the admin page', () => {

    let app, instance;
    
    beforeEach(() => {
        app = shallow(<Admin />);
    });

    it('should match page snapshot', () => {
        expect(app).toMatchSnapshot();
    });

    describe('input validation', () => {

        it('tests program name', () => {
            app.find(TextField).find('#user-program').simulate('change', { target: { value: "program name" } })
            expect(app.state('projectName')).toEqual("program name");
        });

        it('tests file name', () => {
            app.find(TextField).find('#user-file').simulate('change', { target: { value: "file name" } })
            expect(app.state('fileName')).toEqual("file name");
        });

        it('tests github name', () => {
            app.find(TextField).find('#user-github').simulate('change', { target: { value: "github name" } })
            expect(app.state('githubUrl')).toEqual("github name");
        });

        it('tests demo name', () => {
            app.find(TextField).find('#user-demo').simulate('change', { target: { value: "demo name" } })
            expect(app.state('demoUrl')).toEqual("demo name");
        });

        it('tests description name', () => {
            app.find(TextField).find('#user-description').simulate('change', { target: { value: "description name" } })
            expect(app.state('description')).toEqual("description name");
        });

        it('tests description name', () => {
            app.find(TextField).find('#user-description').simulate('change', { target: { value: "description name" } })
            expect(app.state('description')).toEqual("description name");
        });

        it('tests file selected', () => {
            app.find('input').simulate('change', { target: { files: [{ name: "file selected" }] } });
            expect(app.state('fileData')).toEqual({ name: "file selected" });
            expect(app.state('image')).toEqual("file selected");
        });

        
        it('tests valid and invalid submissions', () => {
            app.setState({
                fileName: 'test',
                fileType: '.png',
                fileData: [],
                projectName: 'project',
                githubUrl: 'github',
                demoUrl: 'demo',
                image: 'image',
                description: 'description',
                clicked: false,
            })

            app.find('.submit-button').at(0).simulate('click');
            expect(app.state('clicked')).toBe(true);
        });
    });


    describe('ui interactions', () => {

        let app, instance;

        beforeEach(() => {
            app = mount(<Admin />);
            instance = app.instance();
        })

        afterEach(() => {
            fetchMock.restore();
        });

        it('verifies state changes on `handle submit`', async () => {
            const fetchPromise = Promise.resolve({
                json: () => Promise.resolve(true)
            })

            global.fetch = () => fetchPromise;
            await instance.handleSubmit();
            // expect(app.state('status')).toBe("true");
            // expect(app.state('open')).toBe(true);
        });

        it('verifies state changes on `handle close`', async () => {
            const fetchPromise = Promise.resolve({
                json: () => Promise.resolve(true)
            })

            global.fetch = () => fetchPromise;
            await instance.handleClose();
            expect(app.state('open')).toBe(false);     
        });

    });

});