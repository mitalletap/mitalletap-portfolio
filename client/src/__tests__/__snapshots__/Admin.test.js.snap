// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests the admin page should match page snapshot 1`] = `ShallowWrapper {}`;
